### ISSUE TITLE
CookieAuthenticationHandler is added twice during startup

### ISSUE THREAD

**Issue by Jehoel** [IdentityServer/IdentityServer4#2756]
Please only use the issue tracker for bug reports and/or feature requests. For general security questions, or free or commercial support options do __not__ use the issue tracker and instead see [here](https://identityserver4.readthedocs.io/en/release/intro/support.html) for more details.

For bug reports,  include the relevant log files related to your issue. See here how to enable [logging](https://identityserver4.readthedocs.io/en/release/topics/logging.html). Delete this line once you have.

Finally, please keep the issue concise and to the point. If you paste in more code than the text for the issue you are reporting then we will most likely not read it. 

### Issue / Steps to reproduce the problem

I have this code to configure IdentityServer4:

	var cookiesBefore = services .Where( s => s.ServiceType == typeof(Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler) ).ToList();

	services.AddSingleton<IPersistedGrantStore,MyPersistedGrantStore>();

	services.AddIdentityServer( o =>
		{
			o.Events.RaiseErrorEvents       = true;
			o.Events.RaiseFailureEvents     = true;
			o.Events.RaiseInformationEvents = true;
			o.Events.RaiseSuccessEvents     = true;

			// http://docs.identityserver.io/en/release/topics/signin.html
			o.UserInteraction.ErrorUrl  = "/error"; // The defaults is 'home/error'
			o.UserInteraction.LoginUrl  = "/login"; // Default is 'account/login'
			o.UserInteraction.LogoutUrl = "/logout"; // Default is 'account/logout'
		} )
		.AddSigningCredential( signingCertificate )
		.AddClientStore<MyClientStore>()
		.AddResourceStore<MyResourceStore>()
		.AddProfileService<MyProfileService>()
		.AddResourceOwnerValidator<MyResourceOwnerPasswordValidator>()
		.AddJwtBearerClientAuthentication(); // TODO: How does this compare to ` services.AddAuthentication().AddJwtBearer()`?

	var cookiesAfter = services .Where( s => s.ServiceType == typeof(Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler) ).ToList();

### Relevant parts of the log file

```
Hosting environment: Development
Content root path: C:\git\me\My.Identity\My.Identity.Service
Now listening on: http://localhost:7000
Application started. Press Ctrl+C to shut down.
[14:44:58 Debug] Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler
AuthenticationScheme: idsrv was successfully authenticated.

[14:44:58 Debug] Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler
AuthenticationScheme: idsrv was successfully authenticated.
```

### Problem description

While looking at the output Serilog, I noticed that events from `CookieAuthenticationHandler` were always appearing in pairs. It looks like `CookieAuthenticationHandler` is added to the ASP.NET Core pipeline twice.

I tested this by dumping all instances of `CookieAuthenticationHandler` before and after the call to `AddIdentityServer` and my `cookiesBefore` list contains zero items and `cookiesAfter` contains *two* instances of `CookieAuthenticationHandler`.

**Comment by brockallen** [comment]
I'm going to guess that you have both UseAuthentication and UseIdentityServer in your pipeline.

**Comment by Jehoel** [comment]
> I'm going to guess that you have both UseAuthentication and UseIdentityServer in your pipeline.

I do not. Here is my Configure method:

	public void Configure(IApplicationBuilder app, IHostingEnvironment env)
	{
		MyAutoMapper.Initialize();

		if( env.IsDevelopment() )
		{
			app.UseDeveloperExceptionPage();
		}

		app.UseIdentityServer(); // Invokes .UseAuthentication() internally.

		// https://docs.microsoft.com/en-us/aspnet/core/tutorials/getting-started-with-nswag?view=aspnetcore-2.1&tabs=visual-studio%2Cvisual-studio-xml
		// https://github.com/RSuter/NSwag/issues/1282
		app.UseSwaggerUi3WithApiExplorer( s =>
		{
			s.GeneratorSettings.DefaultPropertyNameHandling = PropertyNameHandling.CamelCase;
		} );

		app.UseMvc( o =>
		{
			// Empty route now set in WelcomeController
			//o.MapRoute( "Default", @"", new { Controller = "Welcome", Action = "Index" } )	
		} );

		app.UseStaticFiles();
	}

**Comment by brockallen** [comment]
Ok, so it's not the same handler is called twice, it's that you have two of the same one, you think?

**Comment by Jehoel** [comment]
As far as I can tell, the `CookieAuthenticationHandler` is called twice during every request that hits my IdentityServer application.

I made an implementation of `IServiceCollection` that logged whenever `CookieAuthenticationHandler` was added and logged the stack-trace when it was added. Both stack-traces are this:

    MyService.exe!
    MyService.MyServiceCollection.Add( Microsoft.Extensions.DependencyInjection.ServiceDescriptor item ) Line 71	C#
    
    Microsoft.Extensions.DependencyInjection.Abstractions.dll! 
    Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddTransient( Microsoft.Extensions.DependencyInjection.IServiceCollection services, System.Type serviceType, System.Type implementationType )	Unknown
    
    Microsoft.AspNetCore.Authentication.dll!
    Microsoft.AspNetCore.Authentication.AuthenticationBuilder.AddSchemeHelper< Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions,  Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler >( string authenticationScheme, string displayName, System.Action<Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions > configureOptions)	Unknown
 	
    IdentityServer4.dll!
    Microsoft.Extensions.DependencyInjection.IdentityServerBuilderExtensionsCore.AddCookieAuthentication(Microsoft.Extensions.DependencyInjection.IIdentityServerBuilder builder)	Unknown
 	
    IdentityServer4.dll!
    Microsoft.Extensions.DependencyInjection.IdentityServerServiceCollectionExtensions.AddIdentityServer(Microsoft.Extensions.DependencyInjection.IServiceCollection services)	Unknown
 	
    Rss.Identity.Service.exe!
    Rss.Identity.Service.Startup.ConfigureIdentityServer(Microsoft.Extensions.DependencyInjection.IServiceCollection services, System.Security.Cryptography.X509Certificates.X509Certificate2 signingCertificate) Line 132	C#

Something fishy is happening in `AddSchemeHelper`.

**Comment by brockallen** [comment]
So I'm not 100% convinced it's the same scheme added twice -- IIRC Microsoft throws if you try that. That's why I assumed at first that the one scheme was just getting called twice (and thus my guess about the authN middleware in there twice).

**Comment by Jehoel** [comment]
I'm looking at the source for `IdentityServerBuilderExtensionsCore.AddCookieAuthentication` and I see that it calls  `AddCookie` twice. I guess the stack-trace removed the stack-frame for `AddCookie` because it was optimized away (I'm running a Release build of IdentityServer4):

    builder.Services.AddAuthentication(IdentityServerConstants.DefaultCookieAuthenticationScheme)
                .AddCookie(IdentityServerConstants.DefaultCookieAuthenticationScheme)
                .AddCookie(IdentityServerConstants.ExternalCookieAuthenticationScheme);

This call is via `AddIdentityServer`:

    public static IIdentityServerBuilder AddIdentityServer(this IServiceCollection services)
        {
            var builder = services.AddIdentityServerBuilder();

            builder
                .AddRequiredPlatformServices()
                .AddCookieAuthentication()                  // <-- here
                .AddCoreServices()
                .AddDefaultEndpoints()
                .AddPluggableServices()
                .AddValidators()
                .AddResponseGenerators()
                .AddDefaultSecretParsers()
                .AddDefaultSecretValidators();

So this looks by-design then - but it still doesn't seem right. In this version 2.2.0 it will always add `AddCookie` even though I'm not using External Authentication.

**Comment by Jehoel** [comment]
Correct, one `CookieAuthenticationHandler` is for `idsrv` the other is for `idsrv.external`.

So why do I get log entries from `CookieAuthenticationHandler` twice about `idsrv`? Surely I'd expect to see one from `idsrv` and the second one from `idsrv.external`?

**Comment by brockallen** [comment]
Again, not sure. Any update?

**Comment by Jehoel** [comment]
I haven't done a deep-dive yet, sorry. I'll need to add some custom logging or maybe build ASP.NET myself from source so I can step-into it and see what's going on, which might take me a while to fit into my calendar.

**Comment by Jehoel** [comment]
On a related note, I'd like to make changes to the default `Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions` (in particular the `OnRedirectToAccessDenied` property. I added `services.PostConfigure` and `services.PostConfigureAll` callbacks but they are never called.

**Comment by brockallen** [comment]
> But when I do this, I get an exception immediately after startup with the message "Scheme already exists: idsrv"

Use a different scheme name, use that in your login pages, and ensure it's the default authentication scheme in ASP.NET Core's AddAuthentication.

**Comment by brockallen** [comment]
Are we all[ sorted out on this issue now, such that we can close?
